File: 2021/02_Dive/dive.py
# ======================================================================
# Dive
#   Advent of Code 2021 Day 02 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         d i v e . p y
# ======================================================================
"A solver for the Advent of Code 2021 Day 02 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                   Dive
# ======================================================================


class Dive(object):   # pylint: disable=R0902, R0205
    "Object for Dive"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.hor = 0
        self.depth = 0
        self.aim = 0

    def one_inst(self, inst):
        "Execute a single instruction"

        # 1. Decode the instruction
        action, number = inst.split()
        number = int(number)

        # 2. Execute the instruction
        if action == "forward":
            self.hor += number
            if self.part2:
                self.depth += self.aim * number
        elif action == "down":
            if self.part2:
                self.aim += number
            else:
                self.depth += number
        elif action == "up":
            if self.part2:
                self.aim -= number
            else:
                self.depth -= number

        # 3. Return pos * depth
        return self.hor * self.depth

    def all_inst(self):
        "Execute all instructions"

        # 1. Loop for all instructions
        for inst in self.text:

            # 2. Execute the instruction
            result = self.one_inst(inst)

        # 3. Return the result
        return result

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.all_inst()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.all_inst()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                          d i v e . p y                         end
# ======================================================================

File: 2021/02_Dive/dive.php
<?php
$lines = file("./input.txt", FILE_IGNORE_NEW_LINES+FILE_SKIP_EMPTY_LINES);
//$lines = array("forward 5", "down 5", "forward 8", "up 3", "down 8", "forward 2");
$position = 0;
$depth = 0;
foreach ($lines as $l) {
    $cmd = explode(" ", $l);
    $num = intval($cmd[1]);
    if ($cmd[0] == 'forward') {
        $position = $position + $num;
    } elseif ($cmd[0] == 'down') {
        $depth = $depth + $num;
    } elseif ($cmd[0] == 'up') {
        $depth = $depth - $num;
    } else {
        echo "unknown command ", $l;
    }
}
$answer = $position * $depth;
echo "<p>Answer to Advent of Code 2021 day 2 part 1 is $answer.</p>";

$position = 0;
$depth = 0;
$aim = 0;
foreach ($lines as $l) {
    $cmd = explode(" ", $l);
    $num = intval($cmd[1]);
    if ($cmd[0] == 'forward') {
        $position = $position + $num;
        $depth = $depth + $aim * $num;
    } elseif ($cmd[0] == 'down') {
        $aim = $aim + $num;
    } elseif ($cmd[0] == 'up') {
        $aim = $aim - $num;
    } else {
        echo "unknown command ", $l;
    }
}
$answer = $position * $depth;
echo "<p>Answer to Advent of Code 2021 day 2 part 2 is $answer.</p>";
?>

